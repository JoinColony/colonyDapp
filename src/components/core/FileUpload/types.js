/* @flow */

export type FileReaderFile = {
  name: string,
  type: string,
  size: number,
  lastModified: string,
  uploadDate: Date,
  data: string,
};

export type UploadFile = {
  file: File,
  uploaded?: string,
  error?: string,
  preview?: string,
};
