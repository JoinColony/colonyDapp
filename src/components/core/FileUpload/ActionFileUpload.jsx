/* @flow */

import React, { Component } from 'react';

import promiseListener from '../../../createPromiseListener';
import FileUpload from './FileUpload.jsx';

import type { AsyncFunction } from '../../../createPromiseListener';

// TODO if this object is sealed, there are unspecified props being used
type Props = {
  submit: string,
  success: string,
  error: string,
};

class ActionFileUpload extends Component<Props> {
  upload: AsyncFunction<void, void>;

  constructor(props: Props) {
    super(props);
    const { submit, success, error } = this.props;
    this.upload = promiseListener.createAsyncFunction({
      start: submit,
      resolve: success,
      reject: error,
    });
  }

  componentWillUnmount() {
    this.upload.unsubscribe();
  }

  render() {
    const { submit, success, error, ...props } = this.props;
    return <FileUpload upload={this.upload.asyncFunction} {...props} />;
  }
}

export default ActionFileUpload;
