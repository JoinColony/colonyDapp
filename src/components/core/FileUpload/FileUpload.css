.main {
  composes: field from '~styles/fields.css';
}

.dropzone {
  padding: 35px;
  border-top: 1px dashed var(--secondary);
  border-right: 2px solid var(--grey-blue-0);
  border-bottom: 1px dashed var(--secondary);
  border-left: 2px solid var(--grey-blue-0);
  background-color: var(--grey-blue-1);
}

.dropzoneAccept {
  background-color: var(--grey-blue-2);
}

.dropzoneReject {
  border: 1px solid var(--danger);
  background-color: color-mod(var(--danger) alpha(30%));
}

.dropzone[aria-invalid="true"] {
  border: 1px solid var(--danger);
  background-color: color-mod(var(--danger) alpha(20%));
}

.placeholderText {
  font-weight: var(--weight-semi);
  text-align: center;
}

.browseButton {
  padding: 0;
  border: none;
  background: transparent;
  font-size: inherit;
  font-weight: inherit;
  color: var(--sky-blue);
  cursor: pointer;
}
