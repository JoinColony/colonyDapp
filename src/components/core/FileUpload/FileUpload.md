### FileUpload

```js
const { Formik } = require('formik');
<Formik
  onSubmit={values => console.log(values)}
  render={
    ({ handleSubmit, isValid }) => (
      <form onSubmit={handleSubmit}>
        <FileUpload
          accept={['application/json',]}
          label="File upload with Formik"
          name="fileUploadFormik"
        />
        <Button appearance={{ theme: 'primary' }} disabled={!isValid} type="submit">And press me</Button>
      </form>
    )
  }
/>
```

### FileUpload with `extra` node

```js
const { Formik } = require('formik');
<Formik
  onSubmit={values => console.log(values)}
  render={
    ({ handleSubmit, isValid }) => (
      <form onSubmit={handleSubmit}>
        <FileUpload
          accept={['application/json',]}
          label="File upload"
          name="fileUpload"
          extra={<span>Wink! Wink!</span>}
        />
        <Button appearance={{ theme: 'primary' }} disabled={!isValid} type="submit">Upload everything!</Button>
      </form>
    )
  }
/>
```
