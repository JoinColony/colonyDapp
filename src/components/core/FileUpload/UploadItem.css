.uploadItem {
  display: flex;
  justify-content: space-between;
}

.fileInfo {
  display: flex;
  flex: 1;
}

.itemProgress {
  flex: 1;
  margin-top: -3px; /* Shift progress bar up to align with text */
  padding: 0 5px;
}

.itemIcon {
  margin-right: 5px;

  & svg {
    stroke: var(--sky-blue);
  }
}

.uploadItem[aria-invalid="true"] {
  & .itemIcon {
    & svg {
      stroke: var(--danger);
    }
  }
}
