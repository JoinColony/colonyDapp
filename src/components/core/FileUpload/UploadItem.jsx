/* @flow */

import React, { Component } from 'react';
import { defineMessages } from 'react-intl';

import fileReader from '../../../lib/fileReader';

import type { UploadFile, FileReaderFile } from './types';

import { asField } from '../Fields';
import { Tooltip } from '../Popover';
import Button from '../Button';
import Icon from '../Icon';
import ProgressBar from '../ProgressBar';

import styles from './UploadItem.css';

const MSG = defineMessages({
  removeActionText: {
    id: 'UploadItem.removeActionText',
    defaultMessage: 'Remove',
  },
  uploadError: {
    id: 'UploadItem.uploadError',
    defaultMessage: 'There was an error uploading your file',
  },
  filetypeError: {
    id: 'UploadItem.filetypeError',
    defaultMessage: 'This filetype is not allowed or file is too big',
  },
});

type Props = {|
  /** Array of allowed file types */
  accept?: string[],
  /** Index of file in list of files to be uploaded */
  idx: number,
  /** Maximum size of file in bytes */
  maxFileSize?: number,
  /** Function used to remove each file from the list of files to upload */
  remove: (idx: number) => void,
  /** Function used to perform the acutal upload action of the file */
  upload: (file: FileReaderFile) => any,
  /** @ignore Will be injected by `asField` */
  $value: UploadFile,
  /** @ignore Will be injected by `asField` */
  $error?: string,
  /** @ignore Will be injected by `asField` */
  setValue: (val: any) => void,
|};

class UploadItem extends Component<Props> {
  _readFiles: (files: Array<Object>) => Promise<Array<Object>>;

  static displayName = 'UploadItem';

  constructor(props: Props) {
    super(props);
    const { accept, maxFileSize } = props;
    this._readFiles = fileReader({
      maxFilesLimit: 1,
      maxFileSize,
      allowedTypes: accept,
    });
  }

  componentDidMount() {
    const {
      $value: { error, file, uploaded },
    } = this.props;

    if (file && !error && !uploaded) {
      this.uploadFile();
    }
  }

  handleRemoveClick = (evt: SyntheticEvent<HTMLButtonElement>) => {
    const { idx, remove } = this.props;
    evt.stopPropagation();
    remove(idx);
  };

  async uploadFile() {
    const { $value, setValue, upload } = this.props;
    let readFile;
    let fileReference;
    const { file } = $value;
    try {
      readFile = await this.read(file);
      setValue({ ...$value, preview: readFile.data });
      fileReference = await upload(readFile);
    } catch (e) {
      // TODO better error handling here
      setValue({ ...$value, error: 'uploadError' });
      return;
    }
    setValue({ ...$value, preview: readFile.data, uploaded: fileReference });
  }

  read = (file: File) => this._readFiles([file]).then(contents => contents[0]);

  render() {
    const {
      $error,
      $value: { file, uploaded },
    } = this.props;

    return (
      <div className={styles.uploadItem} aria-invalid={!!$error}>
        <div className={styles.fileInfo}>
          <Tooltip
            placement="left"
            content={$error || null}
            trigger={$error ? 'hover' : 'disabled'}
          >
            <span className={styles.itemIcon}>
              <Icon name="file" title={file.name} />
            </span>
          </Tooltip>
          {uploaded || $error ? (
            <span className={styles.itemName}>{file.name}</span>
          ) : (
            <div className={styles.itemProgress}>
              <ProgressBar value={uploaded ? 100 : 0} />
            </div>
          )}
        </div>
        <div className={styles.actions}>
          <Button
            type="button"
            onClick={this.handleRemoveClick}
            appearance={{ theme: 'blue' }}
            text={MSG.removeActionText}
          />
        </div>
      </div>
    );
  }
}

const validate = (value: UploadFile) =>
  value.error ? MSG[value.error] : undefined;

export default asField({ alwaysConnected: true, validate })(UploadItem);
