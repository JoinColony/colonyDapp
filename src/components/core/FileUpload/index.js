/* @flow */

export type { FileReaderFile, UploadFile } from './types';

export { default } from './FileUpload.jsx';
export { default as ActionFileUpload } from './ActionFileUpload.jsx';
