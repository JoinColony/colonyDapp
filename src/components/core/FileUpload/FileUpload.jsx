/* @flow */

import type { ComponentType, Node } from 'react';
import type { MessageDescriptor, MessageValues } from 'react-intl';

import React, { Component, Fragment } from 'react';
import Dropzone from 'react-dropzone';
import { getIn } from 'formik';
import { defineMessages, FormattedMessage } from 'react-intl';

import styles from './FileUpload.css';

import { asFieldArray } from '../Fields';
import InputLabel from '../Fields/InputLabel';
import InputStatus from '../Fields/InputStatus';

import type { FileReaderFile, UploadFile } from './types';

import UploadItem from './UploadItem.jsx';

const MSG = defineMessages({
  dropzoneText: {
    id: 'FileUpload.dropzoneText',
    defaultMessage: 'Drag or {browse}',
  },
  labelError: {
    id: 'FileUpload.labelError',
    defaultMessage:
      'There was an error processing your file. Hover over its icon',
  },
  dropzoneTextBrowseAction: {
    id: 'FileUpload.dropzoneTextBrowseAction',
    defaultMessage: 'browse',
  },
});

type ChildProps = {
  isDragActive: boolean,
  isDragAccept: boolean,
  isDragReject: boolean,
};

type Props = {|
  /** Allow specific types of files. See https://github.com/okonet/attr-accept for more information. Keep in mind that mime type determination is not reliable across platforms. CSV files, for example, are reported as text/plain under macOS but as application/vnd.ms-excel under Windows. In some cases there might not be a mime type set at all. See: https://github.com/react-dropzone/react-dropzone/issues/276 */
  accept?: Array<string>,
  children?: Node | (ChildProps => Node),
  /** Custom classNames for different elements of the component */
  classNames: {|
    main: string,
    dropzone: string,
    dropzoneActive: string,
    dropzoneReject: string,
    filesContainer: string,
  |},
  /** Disable the file selection dialog box opening when clicking anywhere in the dropzone */
  disableClick?: boolean,
  /** Get a reference to dropzone, if needed */
  dropzoneRef?: (ref: Dropzone) => void,
  /** Just render the element without label */
  elementOnly?: boolean,
  /** Standard html ID */
  id?: string,
  /** Component to act as the form field  */
  itemComponent: ComponentType<*>,
  /** Maximum number of files to accept */
  maxFilesLimit: number,
  /** Maximum filesize to accept (per-file) */
  maxFileSize: number,
  /** Function to handle the actual uploading of the file */
  upload: (fileData: FileReaderFile) => string,
  /** Input field name (form variable) */
  name: string,
  /** Status text */
  status?: string | MessageDescriptor,
  /** Help text (will appear next to label text) */
  help?: string | MessageDescriptor,
  /** Values for help text (react-intl interpolation) */
  helpValues?: MessageValues,
  /** Extra node to render on the top right in the label */
  extra?: Node,
  /** Label text */
  label?: string | MessageDescriptor,
  /** Values for label text (react-intl interpolation) */
  labelValues?: MessageValues,
  /** Placeholder element for when no files have been picked yet (renderProp) */
  renderPlaceholder?: ?Node,
  /** @ignore injected by `asFieldArray` */
  form: { [string]: any },
  /** @ignore injected by `asFieldArray` */
  push: (file: UploadFile) => void,
  /** @ignore injected by `asFieldArray` */
  remove: (idx: number) => void,
|};

const Placeholder = () => (
  <div className={styles.placeholderText}>
    <FormattedMessage
      {...MSG.dropzoneText}
      values={{
        browse: (
          <span className={styles.browseButton}>
            <FormattedMessage {...MSG.dropzoneTextBrowseAction} />
          </span>
        ),
      }}
    />
  </div>
);

class FileUpload extends Component<Props> {
  static displayName = 'FileUpload';

  static defaultProps = {
    classNames: styles,
    disableClick: false,
    itemComponent: UploadItem,
    maxFilesLimit: 1,
    maxFileSize: 1024 * 1024,
    renderPlaceholder: <Placeholder />,
  };

  addFiles = (acceptedFiles: Array<File>): void => {
    const {
      form: { values },
      name,
      maxFilesLimit,
      push,
    } = this.props;
    const files = getIn(values, name) || [];
    const countAcceptedFiles = files.filter(file => !file.error);
    const newFiles = Array.from(acceptedFiles).slice(
      0,
      maxFilesLimit - countAcceptedFiles.length,
    );
    newFiles.forEach(file => {
      push({ file });
    });
  };

  addRejectedFiles = (rejectedFiles: Array<Object>): void => {
    const {
      form: { values },
      maxFilesLimit,
      push,
      name,
    } = this.props;
    const files = getIn(values, name) || [];
    rejectedFiles.slice(0, maxFilesLimit - files.length).forEach(file => {
      push({ file, error: 'filetypeError' });
    });
  };

  renderExtraChildren = (childProps: ChildProps) => {
    const { children } = this.props;
    if (!children) return null;
    if (typeof children == 'function') {
      return children(childProps);
    }
    return children;
  };

  render() {
    const {
      accept,
      classNames,
      disableClick,
      elementOnly,
      dropzoneRef,
      form: { values, errors, resetForm },
      help,
      extra,
      helpValues,
      itemComponent: FileUploaderItem,
      id,
      label,
      labelValues,
      maxFileSize,
      maxFilesLimit,
      name,
      renderPlaceholder,
      remove,
      upload,
      status,
    } = this.props;

    const files = getIn(values, name) || [];
    const fileErrors = getIn(errors, name) || [];

    const maxFileLimitNotMet = files.length < maxFilesLimit;
    const hasError = !!fileErrors.length;

    return (
      <div className={classNames.main} id={id}>
        {!elementOnly && label && (
          <InputLabel
            label={label}
            help={help}
            labelValues={labelValues}
            helpValues={helpValues}
            extra={extra}
          />
        )}
        <Dropzone
          accept={accept}
          aria-invalid={hasError}
          className={classNames.dropzone}
          activeClassName={classNames.dropzoneActive}
          rejectClassName={classNames.dropzoneReject}
          onDropAccepted={this.addFiles}
          onDropRejected={this.addRejectedFiles}
          disableClick={disableClick || !maxFileLimitNotMet}
          maxSize={maxFileSize}
          ref={dropzoneRef}
        >
          {childProps => (
            <Fragment>
              {maxFileLimitNotMet && renderPlaceholder}
              {files && files.length > 0 && (
                <div className={classNames.filesContainer}>
                  {files.map(({ file }, idx) => (
                    <FileUploaderItem
                      accept={accept}
                      key={`${file.name}-${file.size}`}
                      idx={idx}
                      name={`${name}.${idx}`}
                      remove={remove}
                      reset={resetForm}
                      upload={upload}
                    />
                  ))}
                </div>
              )}
              {this.renderExtraChildren(childProps)}
            </Fragment>
          )}
        </Dropzone>
        <InputStatus status={status} error={hasError ? MSG.labelError : ''} />
      </div>
    );
  }
}

export default asFieldArray()(FileUpload);
