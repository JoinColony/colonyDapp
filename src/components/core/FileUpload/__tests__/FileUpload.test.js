/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { shallowWithIntl } from 'testutils';

import FileUpload from '../FileUpload.jsx';

describe('FileUpload component', () => {
  test('Renders initial component', () => {
    const wrapper = shallowWithIntl(
      <FileUpload
        name="fileUpload"
        label="Basic file upload"
        help="Some help text"
        maxFilesLimit={3}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
