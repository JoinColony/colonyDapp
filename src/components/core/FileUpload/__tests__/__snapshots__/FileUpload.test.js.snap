// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload component Renders initial component 1`] = `
<C
  name="fileUpload"
  render={[Function]}
/>
`;
